Reflect special `DEFAULT` behavior in ciphers(1)

AloisMahdal opened the pull request:
Actual behavior of DEFAULT is different than currently described.
Rather than actinf as cipher string, DEFAULT cannot be combined using
logical operators, etc.

Fixes #5420.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated

richsalz commented:
@AloisMahdal we need an individual CLA from you (we have a global corporate RedHat one).  Thanks.

AloisMahdal commented:
(iCLA sent.)

Note that this PR does not address directly the fact that `DEFAULTz` is equivalent to `DEFAULT:z`, but it does rule out the assumption that it's treated the same as other "cipher lists". In that sense, it's a bug freeze-in (as @vdukhovni points in #5421).

richsalz commented:
See #5429

richsalz commented:
open/close to kick the CLA-bot.

richsalz commented:
How about this minor wording changes:
*The* cipher list can be prefixed with B<DEFAULT> keyword, which enables *the* default
cipher list *as defined below*.  [[delete, it appears below]Content of this list is determined at compile time and
normally corresponds to B<ALL:!COMPLEMENTOFDEFAULT:!eNULL>.]]  Unlike cipher
string*s*, this prefix may not be combined with other strings using B<+> or
B<-> character.  For example, B<DEFAULT+DES> is not a valid cipher list.

vdukhovni commented:
Pretty close, but <-> is not a "combining" character, it is only special as the first character of a cipherlist element (-RC4) but clearly not in the middle (RC4-SHA).

AloisMahdal commented:
Updated:

 *  moved "default" explanation to next paragraph,
 *  added "the",
 *  removed mention of `-` as combining character,
 *  rebased on master.

richsalz commented on doc/man1/ciphers.pod:
```diff
@@ -156,19 +156,20 @@ The cipher string B<@SECLEVEL=n> can be used at any point to set the security
 level to B<n>, which should be a number between zero and five, inclusive.
 See L<SSL_CTX_set_security_level> for a description of what each level means.
 
+Cipher list can be prefixed with B<DEFAULT> keyword, which enables the default
```
"The cipher list"

richsalz commented on doc/man1/ciphers.pod:
```diff
@@ -156,19 +156,20 @@ The cipher string B<@SECLEVEL=n> can be used at any point to set the security
 See L<SSL_CTX_set_security_level> for a description of what each level means.
 
+Cipher list can be prefixed with B<DEFAULT> keyword, which enables the default
+cipher list as defined below.  Unlike cipher string, this prefix may not be
```
Unlike the cipher string ...

AloisMahdal commented:
Thanks for feedback.  (Need to read up something on usage of "the".)  Updated: addded *the*'s, re-wrapped and rebased on master.

vdukhovni commented on doc/man1/ciphers.pod:
```diff
@@ -156,19 +156,20 @@ The cipher string B<@SECLEVEL=n> can be used at any point to set the security
 level to B<n>, which should be a number between zero and five, inclusive.
 See L<SSL_CTX_set_security_level> for a description of what each level means.
 
+The cipher list can be prefixed with B<DEFAULT> keyword, which enables
```
s/B<DEFAULT>/the B<DEFAULT>/

AloisMahdal replied:
Fixed, thanks.

vdukhovni commented on doc/man1/ciphers.pod:
```diff
@@ -156,19 +156,20 @@ The cipher string B<@SECLEVEL=n> can be used at any point to set the security
 See L<SSL_CTX_set_security_level> for a description of what each level means.
 
+The cipher list can be prefixed with B<DEFAULT> keyword, which enables
+the default cipher list as defined below.  Unlike the cipher string,
```
s/Unlike the cipher string/Unlike other cipher strings/

AloisMahdal replied:
Wouldn't that imply `DEFAULT` is a cipher string?  Because it's not... (at least that's what I'm trying to say)

vdukhovni commented on doc/man1/ciphers.pod:
```diff
@@ -156,19 +156,20 @@ The cipher string B<@SECLEVEL=n> can be used at any point to set the security
+The cipher list can be prefixed with B<DEFAULT> keyword, which enables
+the default cipher list as defined below.  Unlike the cipher string,
+this prefix may not be combined with other strings using B<+> character.
+For example, B<DEFAULT+DES> is not a valid cipher list.
```
s/valid cipher list/valid cipher string/

vdukhovni replied:
Mind you, the real story is more subtle (back to the BUGS issue), since `DEFAULT+FOO` is surprisingly instead "DEFAULT:+FOO", which would put FOO last (for any value of FOO). So the users who specify "DEFAULT+FOO" do not get anything like what they probably intended.  If that's described, then it goes into a BUGS section, and must not be described as a stable interface.

AloisMahdal replied:
Fixed the *the*.  See below for comment on BUGS.

vdukhovni commented on doc/man1/ciphers.pod:
```diff
@@ -156,19 +156,20 @@ The cipher string B<@SECLEVEL=n> can be used at any point to set the security
+this prefix may not be combined with other strings using B<+> character.
+For example, B<DEFAULT+DES> is not a valid cipher list.
+
+Content of the default list is determined at compile time and normally
```
s/Content/The content/

AloisMahdal replied:
Fixed, thanks.

AloisMahdal commented:
Fixed most issues (i was not sure about the *the* vs. *other*.

Anyway, I agree with @vdukhovni that it's not a good idea to describe this as stable interface.  I opened (yet) another PR where I described it as part of BUGS section (I added it following convention described in man-pages(7).)

vdukhovni commented on doc/man1/ciphers.pod:
```diff
@@ -156,19 +156,20 @@ The cipher string B<@SECLEVEL=n> can be used at any point to set the security
 See L<SSL_CTX_set_security_level> for a description of what each level means.
 
+The cipher list can be prefixed with the B<DEFAULT> keyword, which enables
+the default cipher list as defined below.  Unlike the cipher string,
```
The text here really does need to change. The simplest change is from 'Unlike the' to 'Unlike other'. If you want to provide more detail, you could explain that all other cipher strings are "+" separated conjunctions (AND) of basic cipher strings, which represent a single cipher or class of ciphers sharing a common elementary property. The "DEFAULT" keywork is not such a cipher string, it is instead a built-in cipher list definition that can be a starting point for further refinement, but is not combinable with "+", and must come first.

levitte replied:
Re s/Unlike the/Unlike other/, the way things are expressed here, `DEFAULT` isn't a cipher string, but rather a prefix.  Seen that way, 'other' would be wrong, wouldn't it?  Although, I might want to see that changed to 'Unlike cipher strings, ...'

richsalz commented:
@AloisMahdal , @vdukhovni any update?

vdukhovni commented:
Just English grammar nits in the docs.

AloisMahdal commented:
Actually, based on @vdukhovni's feedback I've opened also #5455, which takes another approach to solve this #5420 issue by describing the behavior in BUGS section.

My hope is that the community will choose which approach is better and merge one PR and close the other.

levitte commented:
It seems that this is preferred to #5455.  Please see @vdukhovni's commentary there to do some adjustments here.

levitte commented:
This PR seems to have gone stale.  I could see this being approved with the following editorial fixup:

``` diff
diff --git a/doc/man1/ciphers.pod b/doc/man1/ciphers.pod
index 6da316db89..751518d1c2 100644
--- a/doc/man1/ciphers.pod
+++ b/doc/man1/ciphers.pod
@@ -157,9 +157,9 @@ level to B<n>, which should be a number between zero and five, inclusive.
 See L<SSL_CTX_set_security_level> for a description of what each level means.
 
 The cipher list can be prefixed with the B<DEFAULT> keyword, which enables
-the default cipher list as defined below.  Unlike the cipher string,
+the default cipher list as defined below.  Unlike cipher strings,
 this prefix may not be combined with other strings using B<+> character.
-For example, B<DEFAULT+DES> is not a valid cipher string.
+For example, B<DEFAULT+DES> is not valid.
 
 The content of the default list is determined at compile time and normally
 corresponds to B<ALL:!COMPLEMENTOFDEFAULT:!eNULL>.
```

What say @openssl?

richsalz commented:
Yes, please, merge this with your change.  +1

levitte commented:
Merged into master with mentioned editorial changes.

c190506cd827221a861c3207ae2496ea1ec7a1fd Reflect special `DEFAULT` behavior in ciphers(1)
